### DISCUSSION TITLE
AirGap Vault, an air-gapped wallet focussed on security and usability

### DISCUSSION THREAD

**Discussion by AndreasGassmann** [BlockchainCommons/Gordian-Developer-Community#42] [Show and tell]
Hi all,

I would like to introduce our wallet called "AirGap" (https://airgap.it/) to this group. The AirGap project was started about 4 years ago. Ever since then, we have focussed on developing a secure and air-gapped, but also user friendly, wallet.

Back when we started, there was not a lot going on in the air-gapped wallet space. There were only technical options available, and it was nearly impossible to onboard beginners to those solutions. So our goal was to change that and build a better solution.

Some of the key points:
- 2 apps, available in the App Store (iOS) and Play Store (Android)
- The "AirGap Vault" application holds the private keys and does not require an internet connection
- The "AirGap Wallet" application is the companion app to prepare and broadcast transactions
- Communication between "Wallet" and "Vault" is done via QR codes
- Free, open source and reproducible builds (https://walletscrutiny.com/android/it.airgap.vault/)
- Multi currency support (BTC, ETH, DOT, XTZ, KSM, Cosmos, GRS and AE)
- Easy "onboarding" of new users by allowing them to have a one-device setup (with app switching), with an easy path to upgrade to a more secure 2 device, air-gapped setup
- "Paged" QR codes for bigger data
- SSS for secret backup

And some other interesting things we worked on:
- Linux distro without networking that contains only AirGap Vault (currently not up to date because of lack of interest, but will be revived soon): https://github.com/airgap-it/airgap-distro
- Air-gapped APK signing of our apps with a raspberry pi: https://github.com/airgap-it/airgap-raspberry-apk-signer

If you have any questions about our apps, I'm happy to answer them!

---

I found this "Airgapped-Wallet-Community" a couple of months ago when I was investigating the SLIP-39 standard so we could replace our own SSS implementation. Back then I didn't have much time, so when I saw that SLIP-39 is not BIP-39 roundtrip compatible, I stopped looking. Only now I finally had some time and was able to go through all the research that has been going on here. I have to say, great job! There are some very interesting proposals open, such as SSKR (BCR-0011), UR (BCR-2020-005) and UR Types (BCR-2020-006). I hope we can integrate those standards and combine our efforts going forward.

The main reason I am reaching out right now is because the Bitcoin implementation in our app is a bit outdated. We only support legacy addresses, no mutlisig and no other advanced features. So I'm currently working on bringing it up to speed. Besides implementing features like Segwit and PSBTs, we would also like to be compatible with industry standards. We currently use our own data-serialisation (built on RLP, which is used by Ethereum) and QR paging method, so hopefully this can all be replaced by standards defined in this group.

I have some questions:
- Are you planning to submit some of your proposals as BIPs? Or what is the process of making those proposals "final"?
- Is there a javascript / typescript implementation available for:
  - the latest UR spec?
  - the "LifeHash" visualisations?
  - the SSKR algorithm?
- Are there some best practices available how the multisig information should be shared? I have read that there are some issues with this, where some wallets (eg. Electrum) behave differently than other wallets, so I would like to make sure we do it right
- The name "BlockchainCommons" suggests that this group is chain agnostic. However, so far I have only found references to Bitcoin and all the major players here seem to be mainly Bitcoin wallets. Is this on purpose, or are there simply no other blockchains interested?

Thanks for time and I'm looking forward to hearing from you!

**Comment by ChristopherA** [comment]
@AndreasGassmann wrote:
> I would like to introduce our wallet called "AirGap" (https://airgap.it/) to this group. The AirGap project was started about 4 years ago. Ever since then, we have focussed on developing a secure and air-gapped, but also user friendly, wallet.
> 
> Back when we started, there was not a lot going on in the air-gapped wallet space. There were only technical options available, and it was nearly impossible to onboard beginners to those solutions. So our goal was to change that and build a better solution.
> 
> Some of the key points:
> 
> * 2 apps, available in the App Store (iOS) and Play Store (Android)
> * The "AirGap Vault" application holds the private keys and does not require an internet connection
> * The "AirGap Wallet" application is the companion app to prepare and broadcast transactions
> * Communication between "Wallet" and "Vault" is done via QR codes
> * Free, open source and reproducible builds (https://walletscrutiny.com/android/it.airgap.vault/)
> * Multi currency support (BTC, ETH, DOT, XTZ, KSM, Cosmos, GRS and AE)
> * Easy "onboarding" of new users by allowing them to have a one-device setup (with app switching), with an easy path to upgrade to a more secure 2 device, air-gapped setup
> * "Paged" QR codes for bigger data
> * SSS for secret backup
> 
> And some other interesting things we worked on:
> 
> * Linux distro without networking that contains only AirGap Vault (currently not up to date because of lack of interest, but will be revived soon): https://github.com/airgap-it/airgap-distro
> * Air-gapped APK signing of our apps with a raspberry pi: https://github.com/airgap-it/airgap-raspberry-apk-signer
> 
> If you have any questions about our apps, I'm happy to answer them!

Thanks! Welcome to the community! Your experience will be useful.

> I found this "Airgapped-Wallet-Community" a couple of months ago when I was investigating the SLIP-39 standard so we could replace our own SSS implementation. Back then I didn't have much time, so when I saw that SLIP-39 is not BIP-39 roundtrip compatible, I stopped looking. Only now I finally had some time and was able to go through all the research that has been going on here. I have to say, great job! There are some very interesting proposals open, such as SSKR (BCR-0011), UR (BCR-2020-005) and UR Types (BCR-2020-006). I hope we can integrate those standards and combine our efforts going forward.
> 
> The main reason I am reaching out right now is because the Bitcoin implementation in our app is a bit outdated. We only support legacy addresses, no mutlisig and no other advanced features. So I'm currently working on bringing it up to speed. Besides implementing features like Segwit and PSBTs, we would also like to be compatible with industry standards. We currently use our own data-serialisation (built on RLP, which is used by Ethereum) and QR paging method, so hopefully this can all be replaced by standards defined in this group.

Great — you are here at the right time. These topics exactly what a number of us are focusing on right now.

In particular, our wrappers for LibWally might be of interest to you.

> I have some questions:
> 
> * Are you planning to submit some of your proposals as BIPs? Or what is the process of making those proposals "final"?

We've got a first draft of our "release path" strategy at https://github.com/BlockchainCommons/Community/blob/master/release-path.md

Particular relevant is that once we have at least two independent wallet vendors with some level of interoperability, we should move things from [Research](https://github.com/BlockchainCommons/Research) to [WIP](https://github.com/BlockchainCommons/wips) (aka Wallet Improvement Proposals or Work in Progress) — each WIP item should have a target destination, whether BIP, SLIP, IETF, W3C, etc. I consider this community more of an incubator of specifications, reference implementations, and best practices — not a standards body.

> * Is there a javascript / typescript implementation available for:
>   
>   * the latest UR spec?
>   * the "LifeHash" visualisations?
>   * the SSKR algorithm?

I've not seen javascript for any of these yet —Blockchain Commons tends to do C first, sometimes then Rust or Swift, the others have been contributing rust, java, pythons, go, etc. I'd ask in a separate question in this forum as I vaguely heard on the Signal Chat channel that someone was working on LifeHash in Javascript.

> * Are there some best practices available how the multisig information should be shared? I have read that there are some issues with this, where some wallets (eg. Electrum) behave differently than other wallets, so I would like to make sure we do it right

There has been some discussion here, but also more recently in Bitcoin-Dev list at https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2021-February/018388.html

Blockchain Commons is releasing tomorrow a beta of Gordian Cosigner (keep an eye out!), which supports some of the current practices of the so-called BIP38 (which doesn't exist, isn't consistent, and is somewhat broken).  But I feel we ultimately need an answer more like the above Bitcoin-Dev proposal. However, as approach requires state, and the community has to still work with the stateless 48' approaches currently of of Trezor, Ledger & Cold Card, we need to support that as well.

> * The name "BlockchainCommons" suggests that this group is chain agnostic. However, so far I have only found references to Bitcoin and all the major players here seem to be mainly Bitcoin wallets. Is this on purpose, or are there simply no other blockchains interested?

We periodically have interest from the Ethereum wallet community, but so far we seem to have attracted the more serious bitcoin wallet developers. There also has been some interested from the Decentralized Identity community.

The [#SmartCustody](https://www.SmartCustody.com) book we published last year is more chain agnostic — its best practices for responsible key management is good for most coins. However, the book and best practices have not been updated for multisig.

> Thanks for time and I'm looking forward to hearing from you!

I look forward to collaborating with you!

-- Christopher Allen

**Comment by wolfmcnally** [comment]
@AndreasGassmann My intention with LifeHash is that the C++/C implementation is now the "one implementation to rule them all," so rather than do a separate Javascript implementation, what I would like to see is a WASM target for that code, and in fact all the BC libraries. There is truly no need to "re-implement the wheel" here. 😄

~ Wolf

**Reply by AndreasGassmann** [reply]
That would be great. Do you have a timeline for that? I think we will use the available js implementation of `bc-ur` for our upcoming release, but we will hold off working on the `sskr` integration until the WASM module is available.

**Reply by Overtorment** [reply]
a bit too idealistic. for example, react-native has no way to run wasm, but can work with pure js.
a good set of test vectors would be much more preferable

**Reply by AndreasGassmann** [reply]
Can't react native run wasm? I remember there was an issue in the polkadot-wasm library and there appear to be workarounds (Because we don't use react-native I didn't look into it in detail): https://github.com/polkadot-js/wasm/issues/19

Having a native js/ts implementation would be nice (probably also smaller in size), but I would not be comfortable implementing this myself because of the possible security implications. I'm not a cryptographer and from past security audits that we did I learned that you should just stick with well established standards that handle all the corner cases for you.

**Reply by wolfmcnally** [reply]
@ChristopherA How do you want to prioritize getting LifeHash running under WASM?
